### DISCUSSION TITLE
Topic Modeling LDA

### DISCUSSION THREAD

**Discussion by Talioll** [teamalgoritma/community#519] [Q&A] [answered]
Selamat siang, 

Saya ingin berdiskusi mengenai unsupervised machine learning topic modeling LDA. Saat ini saya sedang mengerjakan projek Topic modeling LDA dari review sebuah aplikasi. Saya menemukan kendala yang mana hal tersebut belum bisa diselesaikan dengan maksimal melalui materi yang saya temukan di internet. Adapun masalah yang saya temui sebagai berikut:

**Nilai theta terdistribusi seimbang ke pada seluruh kategori topik sehingga model tidak dapat menetapkan kepada kategori manakah dokumen tersebut masuk.** (walaupun sebenarnya jika dilakuakan conditional if else bisa saja dokumen dikategorikan ke dalam kategori topik, namun tidak mengaplikasikan best practice dari machine learning)

Adapun yang telah saya lakukan:

1. Telah menggunakan nilai k terbaik melalui nilai rata-rata coherence tertinggi (nilai avg coherence tertinggi sebesar 0.102...) 
2.  Telah menggunakan nilai alpha, iterasi dan burnin terbaik (dilihat melakui nilai coherence tertinggi)

Saya menyadari bahwa nilai r-squared dan log-likelihood dari model memang belum berada di nilai yang betul-betul baik (saya sebelum sepenuhnya mendalami hal ini, namun jika dibandingkan dengan kasus lainnya saya merasa model sy belum berada dinilai maksimal). Beberapa indikasi yang saya temukan sebagai salah satu penyebab kurangnya performa model dalam memprediksi dokumen kedalam sebuah kategori topik:

1.  Terdapat terms dengan frekuensi yang sangat banyak mengakibatkan makna yang bias (hal yg telah saya coba lakukan mengeliminasi terms, dan mencoba melakukan TF-IDF namun performa model menurun)
2.  Terdapat dokumen dengan 1 terms saja. Saya memang mempertahankan dokumen yang memiliki 1 terms saja namun begitu dokumen yang hanya memiliki 1 terms karena memang inti dari teks adalah begitu adanya dan secara subjektif saya anggap memiliki kontribusi terhadap analisis isu terhadap urgensi isu yang sedang dibahas sehingga saya pertahankan. (apakah saya perlu eliminasi dokumen dengan kondisi seperti ini?)

Dengan demikian, bagaimanakah solusi yang dapat memaksimalkan model yang saya gunakan ? Apakah terdapat kesalahan dalam proses pengerjaan model yang telah saya lakukan?

**Comment by kevbow** [comment]
Dear @Talioll,

Terima kasih atas pertanyaan yang disampaikan, pertanyaan Anda sudah kami terima. Kami akan mencoba membantu mencarikan referensi berkaitan dengan pertanyaan ini.

**Comment by divakartika** [comment, marked as answer]
Dear @Talioll ,
Mengingat ini adalah dokumen review, tentunya terdapat review yang baik dan review yang buruk. Agar topik yang dihasilkan lebih spesifik, dapat dibuat model terpisah untuk review-review yang baik dan review yang buruk. Pemisahan dapat dilakukan dengan melihat review kuantitatif (mis. bintang 1-2 = review buruk, bintang 4-5 = review baik).

Untuk mengatur jumlah topik dalam satu dokumen, umumnya dilakukan tuning pada hyperparameter alpha, semakin kecil alpha maka satu dokumen akan memiliki jumlah topik yang lebih sedikit. Selain alpha, umumnya juga dilakukan tuning pada parameter beta, yaitu parameter yang mengatur berapa jumlah kata dalam satu topik. Semakin tinggi beta, semakin banyak kata dalam suatu topik.

Referensi:
- https://www.thoughtvector.io/blog/lda-alpha-and-beta-parameters-the-intuition/
- https://stats.stackexchange.com/questions/37405/natural-interpretation-for-lda-hyperparameters
- https://towardsdatascience.com/lda-topic-modeling-an-explanation-e184c90aadcd

Semoga membantu :)

**Reply by Talioll** [reply]
dear @divakartika,

Terima kasih untuk masukannya, untuk informasi tambahan saya sudah mencoba tunning nilai alpha dan beta (nilai alpha sudah tuning menjadi 0.01, dan nilai beta terbaik saya menggunakan nilai default krn ketika di tunning ke nilai lebih rendah maupun ke nilai tinggi performa model menurun). 

Pemisahan data menjadi sesuai kategori score juga sudah saya lakukan (memisahkan data sesuai score kemudian mencari topik sesuai kategori scorenya), hasilnya menunjukan tema topik yang sama sehingga saya berkesimpulan bahwa isu yang sama terjadi di setiap kategori reviewnya sehingga tidak menunjukan perbedaan signifikan dan masih ditemukan rows dengan nilai theta yang masih terdistribusi seimbang (tidak dapat ditentukan probability tertinggi untuk kategori topik tertentu melalui nilai thetanya). Referensi yang diberikan mbak diva akan saya baca terlebih dahulu, semoga dapat menjadi solusi di kasus yang saya temui.

Terima kasih :)
